packages/flutter_tools/lib/src/base/terminal.dart

```diff
@@ -164,11 +165,14 @@ class AnsiTerminal implements Terminal {
     required Platform platform,
     DateTime? now, // Time used to determine preferredStyle. Defaults to 0001-01-01 00:00.
     bool defaultCliAnimationEnabled = true,
+    ShutdownHooks? shutdownHooks,
   })
     : _stdio = stdio,
       _platform = platform,
       _now = now ?? DateTime(1),
-      _isCliAnimationEnabled = defaultCliAnimationEnabled;
+      _isCliAnimationEnabled = defaultCliAnimationEnabled {
+    shutdownHooks?.addShutdownHook(() { singleCharMode = false; });
```

Comment:
I'm guessing this would affect the case in which we're in the middle of `flutter run ...` but the user sends a SIGINT signal to the tool. Are there any other cases where we were failing to set singleCharMode back to false?

Reply:
yes - this is intended to restore the terminal to its normal state if the tool is killed when single character mode is enabled.